### ISSUE TITLE
Monitor / Debugger: Keep Atari screei displayed / show scan line

### ISSUE THREAD

**Issue by craigmaloney** [atari800/atari800#188]
First of all, thank you for an amazing emulator.

I was wondering if there was a way to keep the Atari's screen on while engaging the monitor? Even better would be to show the current scan line / CRT gun position while displaying the image. Something similar either to how Altirra or Stella work. This would be a huge improvement to my development  workflow and understanding of the Atari system in general.

Thank you!

**Comment by mikrosk** [comment]
That's not so hard to achieve, actually. I guess the interesting question would be how to design it. As you have certainly noticed, right now the UI completely covers the screen (and pauses emulation). Maybe there could be a shortcut not to return immediately to emulation but into some sort of "info screen" about current image.

**Comment by joysfera** [comment]
with SDL2 we should be able to open up the monitor/debug in a separate window.

**Comment by mikrosk** [comment]
If I'm not mistaken, @craigmaloney speaks about _graphics_ debug window (and so is our settings window). So it would essentially mean having independent windows open at the same time which may be not trivial (multithreading).

**Comment by craigmaloney** [comment]
Right now for me (under Ubuntu Linux) whenever I hit F8 to go into the monitor it will close the "CRT" screen. I'm left with the debugger/monitor in a text window. Which is fine, but I would like to have the CRT screen left up if possible. Additionally, having it display the scan line and the CRT position that it's currently processing in some way would be extremely handy, 

Thanks for considering this.

**Comment by groessler** [comment]
On 7/17/23 21:30, Craig Maloney wrote:
>
> Right now for me (under Ubuntu Linux) whenever I hit F8 to go into the 
> monitor it will close the "CRT" screen. I'm left with the 
> debugger/monitor in a text window. Which is fine, but I would like to 
> have the CRT screen left up if possible. Additionally, having it 
> display the scan line and the CRT position that it's currently 
> processing in some way would be extremely handy,
>

The X11 target doesn't close the emulation window when entering the 
debugger. Very nice for debugging.


regards,
chris

**Comment by craigmaloney** [comment]
I've done some testing under Linux with the following patch.

```
diff --git a/src/sdl/main.c b/src/sdl/main.c
index 228fbad6..ae39f2fe 100644
--- a/src/sdl/main.c
+++ b/src/sdl/main.c
@@ -108,7 +108,7 @@ int PLATFORM_Exit(int run_monitor)
           for the time the monitor is active. Also, with SDL_VIDEODRIVER=directx all
           keyboard presses in console would be still fetched by the SDL window after
           leaving the monitor. To avoid the problems, close the video subsystem. */
-       SDL_VIDEO_Exit();
+       /*SDL_VIDEO_Exit();*/
        Log_flushlog();
 
        if (run_monitor) {
@@ -121,10 +121,10 @@ int PLATFORM_Exit(int run_monitor)
                        if (!MONITOR_break_step) /*Do not initialise videomode when stepping through code */
 #endif
                        {
-                               SDL_VIDEO_InitSDL();
+                               /*SDL_VIDEO_InitSDL();*/
                                SDL_INPUT_Restart();
                                /* This call reopens the SDL window. */
-                               VIDEOMODE_Update();
+                               /*VIDEOMODE_Update();*/
                        }
        #ifdef SOUND
                        Sound_Continue();

```

Now, I know I'm not foolish enough to declare this a fix (the comment around `SDL_VIDEO_EXIT()` gives me some pause) but it's definitely in the ballpark of what I'm seeking (namely keeping the SDL window around while i"m poking at things in the monitor).

I tried to make this into something more configurable but failed at my attempts. Would someone be able to carry this forward so it's either a configuration for those of us who want this behavior, or the default on platforms that can handle it without causing additional issues?

Thanks!

**Comment by mikrosk** [comment]
Ah, now I see what you meant (for some reason I thought that there isn't a separate console window).

Frankly, the workaround in https://github.com/atari800/atari800/commit/d83f5004a829ced16cdc6fa0f9ce89173afecf05 introduced by @kr0tki isn't something I would call elegant. There must be a way how to flush the accumulated keyboard events when leaving the monitor.

**Comment by joysfera** [comment]
I'd vote for giving @craigmaloney's patch a try. It could be a quick "fix" for now (until we implement multi-window for Atari800). We just need somebody with Windows to see if it indeed makes the keyboard input in monitor unusable or if it's been solved in the meantime somehow (by SDL changes, for example).

**Comment by mikrosk** [comment]
@joysfera what makes you think that @kr0tki's "fix" (which introduced the changes which @craigmaloney reverts now) was Windows specific?

**Comment by groessler** [comment]
See https://sourceforge.net/p/atari800/mailman/message/33630106/

**Comment by mikrosk** [comment]
Ok, I spent a fair chunk of today (yesterday...) verifying the original claim in d83f500:

> With SDL_VIDEODRIVER=directs keyboard presses in console are still
> fetched by the SDL window. This results in locked Return key when
> the user leaves the monitor.

Altough in https://sourceforge.net/p/atari800/mailman/message/33630106 (four years later) Tomasz mentions a completely different reasons:

> resizing the SDL window while in the debugger would 
> result in a variety of glitches after leaving it

This forced me to investigate how to actually build the window executable. In the end I've settled with [quasi-msys2](https://github.com/HolyBlackCat/quasi-msys2) and:
- atari800 compiled against SDL2 + SDL12-compat: OK
- atari800 compiled against SDL1: OK
- atari800 compiled against SDL1 _and_ run from the console with `set SDL_VIDEODRIVER=directx`: the bug is present (!)

Very interesting, even after 13 years the return bug is still there (the "variety of glitches I wasn't able to reproduce).

However. The reason why I pushed the fix anyway is that we shouldn't really be compiling against SDL1 anymore. https://wiki.libsdl.org/SDL2/FAQUsingSDL confirms that SDL2 uses a completely different (and accelerated) Windows driver, so SDL12-compat takes care of that, too.
